### ISSUE TITLE
Support block support class names and inline styles on the block wrapper and make `blockProps` optional

### ISSUE THREAD

**Issue by ockham** [WordPress/block-hydration-experiments#36]
In #3, @luisherranz wrote

> Block supports (only in the wrapper node, no skip-serialization): I'm serializing the block props (which include the `has-text-color`... class names) because I thought React would remove them, but it doesn't seem to do so. I need to investigate more.

"Serializing the block props" means that they're [passed as a `data-gutenberg-block-props` attribute](https://github.com/WordPress/block-hydration-experiments/blob/4631ad7484d10c5cd122fc3221823c73425fc80f/src/gutenberg-packages/frontend.js#L81-L83), and [rehydrated](https://github.com/WordPress/block-hydration-experiments/blob/4631ad7484d10c5cd122fc3221823c73425fc80f/src/gutenberg-packages/frontend.js#L95) based on that, rather than relying on React to take care of the rehydration.

<details>
<summary>
Diff to disable this "manual" rehydration.
</summary>

```diff
diff --git a/src/gutenberg-packages/frontend.js b/src/gutenberg-packages/frontend.js
index 4446742..e0fbe42 100644
--- a/src/gutenberg-packages/frontend.js
+++ b/src/gutenberg-packages/frontend.js
@@ -92,7 +92,7 @@ class GutenbergBlock extends HTMLElement {
                                <EnvContext.Provider value='frontend'>
                                        <Comp
                                                attributes={attributes}
-                                               blockProps={blockProps}
+                                               //blockProps={blockProps}
                                                suppressHydrationWarning={true}
                                                context={context}
                                        >
```
</details>

During [our pair-programming session today](https://github.com/WordPress/block-hydration-experiments/pull/34#issuecomment-1156699778), we actually found that having props of the same name react to interaction causes React to remove the original block props:

```diff
--- a/src/blocks/block-hydration-experiments-parent/frontend/index.js
+++ b/src/blocks/block-hydration-experiments-parent/frontend/index.js
@@ -9,7 +9,7 @@ const Frontend = (
        const [ counter, setCounter ] = useState( initialCounter );
 
        return (
-               <div {...blockProps}>
+               <div className={ show ? 'shown' : 'hidden' } {...blockProps}>
                        <Title message={message} />
                        <Button handler={() => setShow( !show )} />
                        <button onClick={() => setCounter( counter + 1 )}>{counter}</button>
```

Thus, with both diffs applied:

![broken-hydration](https://user-images.githubusercontent.com/96308/173885868-061c111b-c2bd-4d11-9cd3-a792e140c58e.gif)

**Comment by luisherranz** [comment]
So the solution here could be to pass a `blockProps` (or use the `useBlockProps` hook), but make it optional, only for the case where either the class name or the inline style of the block wrapper is going to be dynamically modified in the frontend.

A block that modifies the class name:

```js
const View = ({ blockProps: { className } }) => {
  const [showBlock, setShowBlock] = useState(true);

  return (
    <div className={`${className} ${ showBlock ? "show" : "hide" }`}>
      ...
    </div>
  )
}
```

A block that modifies the inline style:

```js
const View = ({ blockProps: { style } }) => {
  const [width, setWidth] = useState("100px");

  return (
    <div style={{ ...style, width }}>
      ...
    </div>
  )
}
```

A block that doesn't modify the class name nor the inline style:

```js
const View = () => {
  return (
    <div>
      ...
    </div>
  )
}
```

---

_Whether to use a `blockProps` prop or stick to the `useBlockProps` hook is something we could discuss later in the Gutenberg repo if we finally propose this._

**Comment by ockham** [comment]
@luisherranz and I had a pair-programming session earlier today, where we first tried to implement Luis' [idea](https://github.com/WordPress/block-hydration-experiments/issues/36#issuecomment-1170901898):

```diff
diff --git a/src/blocks/block-hydration-experiments-parent/frontend/index.js b/src/blocks/block-hydration-experiments-parent/frontend/index.js
index 1ab5647..5871eb5 100644
--- a/src/blocks/block-hydration-experiments-parent/frontend/index.js
+++ b/src/blocks/block-hydration-experiments-parent/frontend/index.js
@@ -3,13 +3,13 @@ import Button from '../shared/button';
 import Title from '../shared/title';
 
 const Frontend = (
-       { blockProps, attributes: { counter: initialCounter, message }, children },
+       { blockProps: { className }, attributes: { counter: initialCounter, message }, children },
 ) => {
        const [ show, setShow ] = useState( false );
        const [ counter, setCounter ] = useState( initialCounter );
 
        return (
-               <div {...blockProps}>
+               <div className={`${className} ${ show ? "show" : "hide" }`}>
                        <Title message={message} />
                        <Button handler={() => setShow( !show )} />
                        <button onClick={() => setCounter( counter + 1 )}>{counter}</button>
```

However, this did not work: It would serialize the default value for the wrapping `div`'s `className` (`hide`) and keep it around in the hydrated block, even when toggling `show` state. In other words, the `className` would always contain `hide`, even when showing the child block 🙃 

![block-props-hydration](https://user-images.githubusercontent.com/96308/177208393-02a5b9af-7f70-4605-b423-88ed4634300e.gif)

So instead, we tried a different approach: #41

**Comment by luisherranz** [comment]
Thanks, Bernie 🙂

For reference, the difference between the two attempts was:
- First attempt: We picked the `class` and `style` attributes from the DOM during the hydration phase.
- Second attempt: We serialized the block supports output (`class` and `style`) dynamically in PHP using a `data-` attribute.

The component code didn't change, so [this DX](https://github.com/WordPress/block-hydration-experiments/issues/36#issuecomment-1170901898) still applies. _We haven't made `blockProps` optional yet, though._
